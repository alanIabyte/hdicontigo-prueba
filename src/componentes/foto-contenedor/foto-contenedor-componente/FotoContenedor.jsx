import React, { useState, useEffect } from "react";
import { v4 } from "uuid";
import PropTypes from "prop-types";
import ZoomInIcon from "@material-ui/icons/ZoomIn";
import CameraAltIcon from "@material-ui/icons/CameraAlt";
import CheckCircleIcon from "@material-ui/icons/CheckCircle";
import ErrorIcon from "@material-ui/icons/Error";
import {
  EnvolvedorFotoContenedor,
  ImagenFoto,
  ImagenFotoContenedor,
  ImagenFotoSombra,
  FotoContenedorTextos,
  FotoContenedorTitulo,
  FotoContenedorSubtitulo,
} from "./FotoContenedor.styled";

const FotoContenedor = (props) => {
  const diccionario = {
    texto: "Tomar foto",
    textoExito: "Reemplazar foto",
    textoError1: "Carga fallida, reintentar",
    textoError2: "La foto superó el límite de 3 MB, reintentar con otra",
  };
  const {
    imagen,
    imagenPorDefecto,
    titulo,
    alPresentarError,
    alDarClickEnZoom,
    alCargarseExitosamente,
    seCargoAServidor,
  } = props;
  const [imagenAUtilizar, asignarValorImagenAUtilizar] =
    useState(imagenPorDefecto);
  const [cargando, asignarValorCargando] = useState(imagen);
  const [estatus, asignarValorEstatus] = useState();
  const [textoEstatus, asignarValorTextoEstatus] = useState(diccionario.texto);
  const limiteTamanoArchivo = 3145728;

  const actualizarComponente = () => {
    const stringBase64Imagen = imagen.split("base64,")[1];
    const tamanoArchivo = atob(stringBase64Imagen).length;
    const esDeTamanoAdecuado = tamanoArchivo <= limiteTamanoArchivo;
    if (esDeTamanoAdecuado) {
      asignarValorImagenAUtilizar(imagen);
      asignarValorEstatus(seCargoAServidor ? "exito" : "precarga");
      asignarValorTextoEstatus(diccionario.textoExito);
      alCargarseExitosamente();
    } else {
      asignarValorEstatus("error");
      asignarValorTextoEstatus(diccionario.textoError2);
      alPresentarError();
    }
    asignarValorCargando(false);
  };

  useEffect(() => {
    if (imagen) {
      actualizarComponente();
    }
    return null;
  }, []);

  const alDarClickEnContenedor = () => {
    const { alDarClick } = props;
    alDarClick();
  };

  let icon = null;
  if (estatus === "error") {
    icon = <ErrorIcon />;
  } else if (estatus === "exito") {
    icon = <CheckCircleIcon />;
  }

  return (
    <EnvolvedorFotoContenedor
      estatus={estatus}
      key={v4()}
      onClick={alDarClickEnContenedor}
    >
      {imagenAUtilizar ? (
        <ImagenFotoContenedor>
          <ImagenFoto
            porDefecto={!imagen && imagenPorDefecto}
            src={imagenAUtilizar}
          />
          {imagen && estatus === "exito" ? (
            <ImagenFotoSombra onClick={alDarClickEnZoom}>
              <ZoomInIcon />
            </ImagenFotoSombra>
          ) : null}
        </ImagenFotoContenedor>
      ) : null}
      <FotoContenedorTextos conImagen={imagenAUtilizar}>
        <FotoContenedorTitulo estatus={estatus}>{titulo}</FotoContenedorTitulo>
        {!cargando ? <CameraAltIcon /> : null}
        {!cargando ? (
          <FotoContenedorSubtitulo estatus={estatus}>
            {textoEstatus}
          </FotoContenedorSubtitulo>
        ) : null}
      </FotoContenedorTextos>
      {estatus === "" ? null : icon}
    </EnvolvedorFotoContenedor>
  );
};

FotoContenedor.defaultProps = {
  alCargarseExitosamente: () => null,
  alDarClick: () => null,
  alDarClickEnZoom: () => null,
  alPresentarError: () => null,
  imagen: "",
  imagenPorDefecto: "",
  seCargoAServidor: false,
  titulo: "",
};

FotoContenedor.propTypes = {
  alCargarseExitosamente: PropTypes.func,
  alDarClick: PropTypes.func,
  alDarClickEnZoom: PropTypes.func,
  alPresentarError: PropTypes.func,
  imagen: PropTypes.string,
  imagenPorDefecto: PropTypes.string,
  seCargoAServidor: PropTypes.bool,
  titulo: PropTypes.string,
};

export default FotoContenedor;
