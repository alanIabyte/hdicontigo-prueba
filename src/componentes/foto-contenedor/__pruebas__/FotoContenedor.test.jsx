import React from "react";
import renderer from "react-test-renderer";
import FotoContenedor from "../foto-contenedor-componente/FotoContenedor";
import FotoContenedorSimulacion from "../__simulaciones__/FotoContenedor.simulacion";

describe("FotoContenedor", () => {
  it("se despliega correctamente", () => {
    const { reporte } = FotoContenedorSimulacion;
    const tree = renderer.create(<FotoContenedor reporte={reporte} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
