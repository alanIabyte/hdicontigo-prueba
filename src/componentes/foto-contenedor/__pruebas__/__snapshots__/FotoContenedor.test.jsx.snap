// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FotoContenedor se despliega correctamente 1`] = `
<div
  className="sc-ikkxIA fsHraw"
  onClick={[Function]}
>
  <div
    className="sc-hmdomO jGsBwh"
  >
    <div
      className="sc-iGgWBj sc-dAbbOL fgQoGq jqEAFy"
    >
      
    </div>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <circle
        cx="12"
        cy="12"
        r="3.2"
      />
      <path
        d="M9 2L7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"
      />
    </svg>
    <div
      className="sc-kpDqfm sc-feUZmu bTCkFB bAbFDS"
    >
      Tomar foto
    </div>
  </div>
</div>
`;
